hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/StakingPeriodTimeHook.java

```diff
@@ -0,0 +1,92 @@
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.hedera.node.app.service.token.impl.handlers.staking.EndOfStakingPeriodUpdater;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import com.hedera.node.config.data.StakingConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+@Singleton
+public class StakingPeriodTimeHook implements ConsensusTimeHook {
+    private static final Logger logger = LogManager.getLogger(StakingPeriodTimeHook.class);
+
+    private final EndOfStakingPeriodUpdater stakingCalculator;
+    private Instant consensusTimeOfLastHandledTxn;
+
+    @Inject
+    public StakingPeriodTimeHook(@NonNull final EndOfStakingPeriodUpdater stakingPeriodCalculator) {
+        this.stakingCalculator = requireNonNull(stakingPeriodCalculator);
+    }
+
+    @Override
+    public void process(@NonNull Instant consensusTime, @NonNull final HandleContext context) {
+        if (consensusTimeOfLastHandledTxn == null
+                || consensusTime.getEpochSecond() > consensusTimeOfLastHandledTxn.getEpochSecond()
+                        && isNextPeriod(consensusTime, consensusTimeOfLastHandledTxn, context)) {
+            // Handle the daily staking distributions and updates
+            try {
+                stakingCalculator.updateNodes(consensusTime, context);
+            } catch (final Exception e) {
+                logger.error("CATASTROPHIC failure updating end-of-day stakes", e);
+            }
+
+            // Advance the last consensus time to the given consensus time
+            consensusTimeOfLastHandledTxn = consensusTime;
+        }
+    }
+
+    @VisibleForTesting
+    void setLastConsensusTime(@Nullable final Instant lastConsensusTime) {
+        consensusTimeOfLastHandledTxn = lastConsensusTime;
+    }
+
+    @VisibleForTesting
+    static boolean isNextPeriod(
+            @NonNull final Instant currentConsensusTime,
+            @NonNull final Instant previousConsensusTime,
+            @NonNull final HandleContext handleContext) {
+        final var stakingPeriod =
+                handleContext.configuration().getConfigData(StakingConfig.class).periodMins();
+        if (stakingPeriod == DEFAULT_STAKING_PERIOD_MINS) {
+            return isLaterUtcDay(currentConsensusTime, previousConsensusTime);
+        } else {
+            return getPeriod(currentConsensusTime, stakingPeriod * MINUTES_TO_MILLISECONDS)
+                    != getPeriod(previousConsensusTime, stakingPeriod * MINUTES_TO_MILLISECONDS);
+        }
+    }
+
+    private static boolean isLaterUtcDay(@NonNull final Instant now, @NonNull final Instant then) {
+        final var nowDay = LocalDateTime.ofInstant(now, UTC);
```

Comment:
_Is_ `LocalDate` sufficient? If we want to update immediately after midnight, I would think that going by UTC time is important

Reply:
[LocalDate.ofInstant](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/time/LocalDate.html#ofInstant(java.time.Instant,java.time.ZoneId)) uses the time zone just as LocalDateTime does, so it would handle the UTC midnight margin.
You still need to check dayOfYear (so it doesn't change the calculation significantly), but it does make clearer that you only care about the day...

Reply:
Ah, I was looking at using `LocalDate.from(...)`. Will change to `LocalDate.ofInstant(...)`